google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java

```diff
@@ -192,42 +199,51 @@ public String getTopicNameString() {
       throw new IllegalStateException("Cannot publish on a shut-down publisher.");
     }
 
+    final String orderingKey = message.getOrderingKey();
+    if (!orderingKey.isEmpty() && !enableMessageOrdering) {
+      throw new IllegalStateException(
+          "Cannot publish a message with an ordering key when message ordeirng is not enabled.");
+    }
+
     final int messageSize = message.getSerializedSize();
     OutstandingBatch batchToSend = null;
     SettableApiFuture<String> publishResult = SettableApiFuture.<String>create();
     final OutstandingPublish outstandingPublish = new OutstandingPublish(publishResult, message);
     messagesBatchLock.lock();
```

Comment:
better to use lock per ordering key.

Reply:
There is a comment in `publishAllOutstanding` that says unlocking messagesBatchLock there could result in out-of-order messaging. Even if we have a per-ordering-key lock, we'll still need the global one as well in order to insert into `messageBatches`. I think the move to more fine-grained locking is a performance enhancement we can look into as we run more tests and after a change to avoid premature batching, which will come soon. For now, I want to make sure we get the correctness right and understand the scope of the work being undertaken so we can start it in client libraries in the other languages.